### ISSUE TITLE
Separate Derive and Decoration type

### ISSUE THREAD

**Issue by bogeychan** [elysiajs/elysia#291]
I tried to do [remapping](https://elysiajs.com/blog/elysia-07.html#remapping) with the latest elysia version:

_TypeScript's type suggestion works, but the method does not exist at runtime_

### Code to reproduce

```ts
import { Elysia } from 'elysia';

const plugin = new Elysia().derive(() => ({
  pluginMethod() {
    console.log('pluginMethod');
  }
}));

const app = new Elysia()
  .use(
    plugin.decorate(({ pluginMethod, ...rest }) => ({
      myPluginMethod: pluginMethod,
      ...rest
    }))
  )
  .get('/', (ctx) => {
    ctx.myPluginMethod();
  })
  .listen(8080);

console.log(`http://${app.server!.hostname}:${app.server!.port}`);
```

### Output in browser

```txt
ctx.myPluginMethod is not a function. (In 'ctx.myPluginMethod()', 'ctx.myPluginMethod' is undefined)
```

### Environment

- elysia@0.7.21
- bun@1.0.7

**Comment by SaltyAom** [comment]
Sorry for not getting back to you sooner, I just got out of the hospital.

This doesn't work because `decorate` and `derive` has a separate store, if you wants to remap the `derive` function, you should use `derive` instead.
```ts
import { Elysia } from 'elysia'

const plugin = new Elysia().derive(() => ({
	pluginMethod() {
		console.log('pluginMethod')
	}
}))

const app = new Elysia()
	.use(
		plugin.derive(({ pluginMethod, ...rest }) => ({
			myPluginMethod: pluginMethod,
			...rest
		}))
	)
	.get('/', (ctx) => {
		ctx.myPluginMethod()
	})
	.listen(8080)
```

Which means the code above should just work.

This means that there's a type bug in `derive` function instead of remapping.
Please keep this issue open until the patch arrive.

**Comment by adicco** [comment]
Any update on this? I am experiencing typing issues with `derive` as well

**Comment by SaltyAom** [comment]
Fixed since release of 0.8.0 as mentioned in release note.
> differentiate derive from decorator['request'] as decorator['derive']

See https://elysiajs.com/blog/elysia-08.html#notable-improvement

Unable to reproduce with the following code anymore:
```ts
const plugin = new Elysia()
	.derive(() => ({
		pluginMethod() {
			console.log('pluginMethod')
		}
	}))
	.derive(({ pluginMethod, ...rest }) => ({
		myPluginMethod: pluginMethod,
		...rest
	}))
	.as('plugin')

const app = new Elysia().use(plugin).get('/', (ctx) => {
	ctx.myPluginMethod()
})
```

Closing as complete.
